internal/auth/auth.go

```diff
@@ -522,25 +523,53 @@ func (b *bearerHandler) AddScope(scope string) error {
 		}
 		return nil
 	}
-	return b.addScope(scope)
+	b.addScope(scope)
+	return nil
 }
 
-func (b *bearerHandler) addScope(scope string) error {
-	replaced := false
-	for i, cur := range b.scopes {
-		// extend an existing scope with more actions
-		if strings.HasPrefix(scope, cur+",") {
-			b.scopes[i] = scope
-			replaced = true
-			break
-		}
-	}
-	if !replaced {
+var (
+	scopeRe        = regexp.MustCompile(`(repository:[^:]+:)([^:]+)`)
+	knownActionsRe = regexp.MustCompile(`^(push|pull|delete)(,(push|pull|delete))*$`)
+)
+
+func (b *bearerHandler) addScope(scope string) {
+	scopeMatch := scopeRe.FindStringSubmatch(scope)
+	if len(scopeMatch) < 3 || !b.tryExtendExistingScope(scopeMatch[1], scopeMatch[2]) {
 		b.scopes = append(b.scopes, scope)
 	}
 	// delete old token
 	b.token.Token = ""
-	return nil
+}
+
+func (b *bearerHandler) tryExtendExistingScope(repoPrefix, newActions string) bool {
+	for i, cur := range b.scopes {
+		curMatch := scopeRe.FindStringSubmatch(cur)
+		if len(curMatch) < 3 {
+			continue
+		}
```

Comment:
I'd add a fast failure to the top of the loop, and with `parseScope` on the entries:

```suggestion
	scopePrefix := "repository:" + repo + ":"
	for i, cur := range b.scopes {
		if !strings.HasPrefix(cur, scopePrefix) {
			continue
		}
		curRepo, curActions, ok := parseScope(cur)
		if !ok {
			continue
		}
		// add unique actions to curActions
		for _, a := range actions {
			inList := false
			for _, ca := range curActions {
				if a == ca {
					inList = true
					break
				}
				if !inList {
					curActions = append(curActions, a)
				}
			}
		}
		b.scopes[i] = scopePrefix + strings.Join(curActions, ",")
		return true
	}
```

Reply:
Done, slightly improved last for loop.